testing/deployer/topology/compile_test.go

```diff
@@ -0,0 +1,2282 @@
+							Name:      "mesh1",
+							Addresses: []*Address{{Network: "foo"}},
+							Workloads: []*Workload{{
+								ID:                     NewID("zim", "", ""),
+								Image:                  "busybox",
+								EnvoyAdminPort:         19000,
+								Port:                   8443,
+								DisableServiceMesh:     true,
+								EnableTransparentProxy: true,
+							}},
+						},
+					},
+				}},
+			},
+			expectErr: `error building cluster "foo": error compiling node "mesh1": cluster "foo" node "mesh1" service "default/default/zim" is not valid: cannot disable service mesh and activate tproxy`,
+		},
+		"workload/validate/disable-mesh/set admin port": {
+			in: &Config{
+				Networks: []*Network{
+					{Name: "foo"},
+				},
+				Clusters: []*Cluster{{
+					Name: "foo",
+					Nodes: []*Node{
+						{
+							Kind: NodeKindServer,
+							Name: "server1",
+							Addresses: []*Address{
+								{Network: "foo"},
+							},
+						},
+						{
+							Kind:      NodeKindDataplane,
+							Version:   NodeVersionV2,
+							Name:      "mesh1",
+							Addresses: []*Address{{Network: "foo"}},
+							Workloads: []*Workload{{
+								ID:                 NewID("zim", "", ""),
+								Image:              "busybox",
+								EnvoyAdminPort:     19000,
+								Port:               8443,
+								DisableServiceMesh: true,
+							}},
+						},
+					},
+				}},
+			},
+			expectErr: `error building cluster "foo": error compiling node "mesh1": cluster "foo" node "mesh1" service "default/default/zim" is not valid: cannot use envoy admin port without a service mesh`,
+		},
+		"workload/validate/enable-mesh/unset admin port": {
+			in: &Config{
+				Networks: []*Network{
+					{Name: "foo"},
+				},
+				Clusters: []*Cluster{{
+					Name: "foo",
+					Nodes: []*Node{
+						{
+							Kind: NodeKindServer,
+							Name: "server1",
+							Addresses: []*Address{
+								{Network: "foo"},
+							},
+						},
+						{
+							Kind:      NodeKindDataplane,
+							Version:   NodeVersionV2,
+							Name:      "mesh1",
+							Addresses: []*Address{{Network: "foo"}},
+							Workloads: []*Workload{{
+								ID:    NewID("zim", "", ""),
+								Image: "busybox",
+								Port:  8443,
+							}},
+						},
+					},
+				}},
+			},
+			expectErr: `error building cluster "foo": error compiling node "mesh1": cluster "foo" node "mesh1" service "default/default/zim" is not valid: envoy admin port is required`,
+		},
+
+		// TODO: OK: v2 mesh gateway port defaulting
+
+		// TODO: collect tenancies from all places (cluster, initialconfigs, initialres, nodes, services, ...)
+	}
+
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			run(t, tc)
+		})
+	}
+}
+
+type testingT interface {
+	Helper()
+	Fatalf(format string, args ...interface{})
+	require.TestingT
+}
+
+var ignoreUnexportedTypes = []any{
+	Cluster{},
+	Images{},
+	Node{},
+	Workload{},
+}
+
+func assertDeepEqual[V any](t testingT, exp, got V, msgAndArgs ...any) {
```

Comment:
This is an inlined, modified version of `prototest.AssertDeepEqual` for use in this module.